sematic/calculator.py

```diff
@@ -61,6 +60,62 @@ def __init__(
         self.__doc__ = func.__doc__
         self.__module__ = func.__module__
         self.__name__ = func.__name__
+        self._validate()
+
+    def _validate(self):
+        for key, annotation in self._input_types.items():
+            try:
+                validate_type_annotation(annotation)
+            except TypeError as e:
+                raise TypeError(
+                    f"Invalid type annotation for argument '{key}' of {repr(self)}: {e}"
+                )
+        try:
+            validate_type_annotation(self._output_type)
+        except TypeError as e:
+            raise TypeError(f"Invalid type annotation for output of {repr(self)}: {e}")
+
+    @classmethod
+    def _validate_func(cls, func):
+        is_method_like = (
+            inspect.ismethod(func)
+            or
+            # Why is this second check necessary? Because if somebody uses @func
+            # on a method inside the class itself, the method is not yet bound
+            # to the class, and inspect's ismethod won't pick it up.
+            (
+                hasattr(func, "__call__")
+                and len(inspect.signature(func).parameters.keys()) >= 1
+                and list(inspect.signature(func).parameters.keys())[0]
+                in {"self", "cls"}
+            )
+        )
+        if is_method_like:
+            raise TypeError(
+                "@sematic.func can only be used with functions, not methods. "
+                f"But {repr(func)} is a method. Instead, consider defining a "
+                "function where the first argument is the object currently being "
+                "passed as 'self' or 'cls'."
+            )
+        is_coroutine_like = (
+            inspect.isawaitable(func)
+            or inspect.isasyncgenfunction(func)
+            or inspect.isasyncgen(func)
+            or inspect.iscoroutine(func)
+            or inspect.iscoroutinefunction(func)
+            or inspect.isgenerator(func)
+            or inspect.isgeneratorfunction(func)
+        )
```

Comment:
Readability nit:

```python
is_coroutine_like = any(
    is_(func) for is_ in {
        inspect.isawaitable,
        inspect.isasyncgenfunction,
        inspect.isasyncgen,
        inspect.iscoroutine,
        inspect.iscoroutinefunction,
        inspect.isgenerator,
        inspect.isgeneratorfunction
})
```

Reply:
Also, maybe cut out `is_method_like` and `is_coroutine_like` into standalone module-level private functions.

Reply:
done and done